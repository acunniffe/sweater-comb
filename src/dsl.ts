import {
  ApiCheckDsl,
  EntityRule,
  Result,
  DocsLinkHelper,
  newDocsLinkHelper,
  runCheck,
} from "@useoptic/api-checks";
import {
  ConceptualLocation,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiOperationFact,
  IChange,
  IFact,
  OpenApiFieldFact,
  jsonPointerHelper,
  ILocation,
  OpenAPIV3,
} from "@useoptic/openapi-utilities";
import { genericEntityRuleImpl } from "@useoptic/api-checks/build/sdk/generic-entity-rule-impl";
import { ShouldOrMust } from "@useoptic/api-checks/build/sdk/types";
import { OpenApiResponseFact } from "@useoptic/openapi-utilities/build/openapi3/implementations/openapi3/openapi-traverser";

type SnykStablity = "wip" | "experimental" | "beta" | "ga";
type DateString = string; // YYYY-mm-dd
type ResourceName = string;

export interface SynkApiCheckContext {
  // Vervet provides context about the change itself. Since
  // Optic is analyzing two OpenAPI spec files, we need to tell it
  // when the change is supposed to happen, and the resource/version info
  // determined by the file's location in a directory structure.
  changeDate: DateString; // when the change did (or would, if proposed) occur
  changeResource: ResourceName; // the spec resource being changed
  changeVersion: {
    // the spec version being changed
    date: DateString;
    stability: SnykStablity;
  };

  // Vervet provides a mapping that indicates the resource version deprecation.
  // It has this information because it processes the entire source tree of
  // spec files.
  resourceVersions: {
    [ResourceName: string]: {
      // changeResource used to match this
      [DateString: string]: {
        // changeVersion.date used to match this
        [SnykStablity: string]: {
          // changeVersion.stability matches this
          deprecatedBy: {
            // the spec version that deprecates this one (if any) or null
            date: DateString;
            stability: SnykStablity; // could be higher stability than changed!
          } | null;
        };
      };
    };
  };
}

export interface SnykEntityRule<T, A>
  extends EntityRule<T, ConceptualLocation, SynkApiCheckContext, A> {}

export class SnykApiCheckDsl implements ApiCheckDsl {
  private checks: Promise<Result>[] = [];

  constructor(
    private nextFacts: IFact<any>[],
    private changelog: IChange<any>[],
    private nextJsonLike: OpenAPIV3.Document,
    private providedContext: SynkApiCheckContext
  ) {}

  checkPromises() {
    return this.checks;
  }

  getContext(location: ILocation): ConceptualLocation & SynkApiCheckContext {
    return {
      ...location.conceptualLocation,
      ...this.providedContext,
    };
  }

  get operations() {
    return genericEntityRuleImpl<
      OpenApiOperationFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.OperationObject
    >(
      OpenApiKind.Operation,
      this.changelog,
      this.nextFacts,
      (opFact) => `${opFact.method.toUpperCase()} ${opFact.pathPattern}`,
      (location) => this.getContext(location),
      (...items) => this.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(this.nextJsonLike, pointer)
    );
  }

  get context() {
    const change: IChange<SnykApiCheckDsl> = {
      location: {
        conceptualLocation: { path: "This Specification", method: "" },
        jsonPath: "/",
        conceptualPath: [],
        kind: "API",
      },
    };

    const value: ShouldOrMust<
      (
        context: SynkApiCheckContext,
        docs: DocsLinkHelper
      ) => Promise<void> | void
    > = {
      must: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          true,
          () => handler(this.providedContext, docsHelper)
        );
      },
      should: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          false,
          () => handler(this.providedContext, docsHelper)
        );
      },
    };

    return value;
  }

  get responses() {
    const dsl = this;

    return {
      ...genericEntityRuleImpl<
        OpenApiResponseFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.ResponsesObject
      >(
        OpenApiKind.Response,
        dsl.changelog,
        dsl.nextFacts,
        (response) => `response ${response.statusCode}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
      headers: genericEntityRuleImpl<
        OpenApiHeaderFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.HeaderObject
      >(
        OpenApiKind.ResponseHeader,
        dsl.changelog,
        dsl.nextFacts,
        (header) => `response header ${header.name}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
    };
  }

  get checkApiContext(): ShouldOrMust<
    (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
  > {
    const contextChangedHandler: (must: boolean) => ContextChangedRule["must"] =
      (must: boolean) => {
        return (statement, handler) => {
          const docsHelper = newDocsLinkHelper();
          const syntheticChange = {
            added: this.providedContext,
            location: {
              jsonPath: "/",
              conceptualPath: [],
              conceptualLocation: {
                path: "Entire Resource",
                method: "",
              },
              kind: "ContextRule",
            },
          };
          return runCheck(
            syntheticChange,
            docsHelper,
            "api lifeycle: ",
            statement,
            must,
            () => handler(this.providedContext, docsHelper)
          );
        };
      };

    return {
      must: contextChangedHandler(true),
      should: contextChangedHandler(false),
    };
  }

  get bodyProperties(): SnykEntityRule<
    OpenApiFieldFact,
    OpenAPIV3.SchemaObject
  > {
    const dsl = this;
    return genericEntityRuleImpl<
      OpenApiFieldFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.SchemaObject
    >(
      OpenApiKind.Field,
      dsl.changelog,
      dsl.nextFacts,
      (field) => `field ${field.key}`,
      (location) => dsl.getContext(location),
      (...items) => dsl.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
    );
  }
}

type ContextChangedRule = ShouldOrMust<
  (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
>;
