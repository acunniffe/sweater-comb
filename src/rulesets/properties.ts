import { SnykApiCheckDsl } from "../dsl";
const { expect } = require("chai");
export const rules = {
  propertyKey: ({ bodyProperties }: SnykApiCheckDsl) => {
    bodyProperties.requirement.must("have camel case keys", ({ key }) => {
      const snakeCase = /^[a-z]+(?:_[a-z]+)*$/g;
      expect(snakeCase.test(key)).to.be.ok;
    });
  },
  propertyExample: ({ bodyProperties }: SnykApiCheckDsl) => {
    bodyProperties.requirement.must("have an example", ({ flatSchema }) => {
      expect(flatSchema.example).to.exist;
    });
  },
  propertyFormat: ({ bodyProperties }: SnykApiCheckDsl) => {
    bodyProperties.requirement.should(
      "have a format when a string",
      ({ flatSchema }) => {
        if (flatSchema.type === "string") {
          expect(flatSchema.format).to.exist;
        }
      }
    );
  },
  preventRemoval: ({ bodyProperties }: SnykApiCheckDsl) => {
    bodyProperties.removed.must("not be removed", (property) => {
      expect(false, `expected ${property.key} to be present`).to.be.ok;
    });
  },
  preventAddingRequiredRequestProperties: ({
    bodyProperties,
  }: SnykApiCheckDsl) => {
    bodyProperties.added.must("not be required", (property, context) => {
      if (!context.inRequest) return;
      expect(property.required).to.not.be.true;
    });
  },
};
