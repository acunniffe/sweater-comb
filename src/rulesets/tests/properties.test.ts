import { rules } from "../properties";
import { createTestDslFixture } from "@useoptic/api-checks";
import { SnykApiCheckDsl, SynkApiCheckContext } from "../../dsl";

// todo: fix copy/paste
const { compare } = createTestDslFixture<SnykApiCheckDsl, SynkApiCheckContext>(
  (input) => {
    return new SnykApiCheckDsl(
      input.nextFacts,
      input.changelog,
      input.nextJsonLike,
      input.context
    );
  }
);

// todo: fix copy/paste
const emptyContext: SynkApiCheckContext = {
  changeDate: "2021-10-10",
  changeResource: "Example",
  changeVersion: {
    date: "2021-10-10",
    stability: "ga",
  },
  resourceVersions: {},
};

describe("body properties", () => {
  const baseOpenAPI = {
    openapi: "3.0.1",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "Empty" },
  };

  describe("key", () => {
    it("passes when camel case", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      isCamelCase: { type: "string" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyKey, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("fails when not camel case", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      "not-camel-case": { type: "string" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyKey, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("example", () => {
    it("passes if exists", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      name: { type: "string", example: "Jane Doe" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyExample, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("fails if doesn't exist", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      name: { type: "string" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyExample, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("format", () => {
    it("fails if format doesn't exist", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      name: { type: "string" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyFormat, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result.results[0].isShould).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("passes if not a string", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              content: {
                "application/json": {
                  schema: {
                    type: "object",
                    properties: {
                      count: { type: "number" },
                    },
                  },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.propertyFormat, emptyContext);

      expect(result.results[0].isShould).toBeTruthy();
      expect(result).toMatchSnapshot();
    });
  });
});
