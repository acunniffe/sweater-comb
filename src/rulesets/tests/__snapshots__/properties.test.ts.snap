// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`body properties example fails if doesn't exist 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "type": "string",
        },
        "key": "name",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "name",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "name",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "name": Object {
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "name",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "name",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "name",
          "required": false,
        },
      },
      "condition": "have an example",
      "docsLink": undefined,
      "error": "expected undefined to exist",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for field: field name",
    },
  ],
}
`;

exports[`body properties example passes if exists 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "example": "Jane Doe",
          "type": "string",
        },
        "key": "name",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "name",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "name",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "name": Object {
                        "example": "Jane Doe",
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "name",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "name",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "example": "Jane Doe",
            "type": "string",
          },
          "key": "name",
          "required": false,
        },
      },
      "condition": "have an example",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for field: field name",
    },
  ],
}
`;

exports[`body properties format fails if format doesn't exist 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "type": "string",
        },
        "key": "name",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "name",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "name",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "name": Object {
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "name",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "name",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/name",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "name",
          "required": false,
        },
      },
      "condition": "have a format when a string",
      "docsLink": undefined,
      "error": "expected undefined to exist",
      "isMust": false,
      "isShould": true,
      "passed": false,
      "where": "requirement for field: field name",
    },
  ],
}
`;

exports[`body properties format passes if not a string 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "type": "number",
        },
        "key": "count",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "count",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "count",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/count",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "count": Object {
                        "type": "number",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "count",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "count",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/count",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "number",
          },
          "key": "count",
          "required": false,
        },
      },
      "condition": "have a format when a string",
      "docsLink": undefined,
      "isMust": false,
      "isShould": true,
      "passed": true,
      "where": "requirement for field: field count",
    },
  ],
}
`;

exports[`body properties key fails when not camel case 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "type": "string",
        },
        "key": "not-camel-case",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "not-camel-case",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "not-camel-case",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/not-camel-case",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "not-camel-case": Object {
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "not-camel-case",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "not-camel-case",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/not-camel-case",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "not-camel-case",
          "required": false,
        },
      },
      "condition": "have camel case keys",
      "docsLink": undefined,
      "error": "expected false to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for field: field not-camel-case",
    },
  ],
}
`;

exports[`body properties key passes when camel case 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "contentType": "application/json",
        "flatSchema": Object {
          "type": "object",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json",
        "kind": "body",
      },
    },
    Object {
      "added": Object {
        "flatSchema": Object {
          "type": "string",
        },
        "key": "isCamelCase",
        "required": false,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "body": Object {
              "contentType": "application/json",
            },
            "statusCode": "200",
          },
          "jsonSchemaTrail": Array [
            "isCamelCase",
          ],
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "application/json",
          "isCamelCase",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/isCamelCase",
        "kind": "field",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "isCamelCase": Object {
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "isCamelCase",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "isCamelCase",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/isCamelCase",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "isCamelCase",
          "required": false,
        },
      },
      "condition": "have camel case keys",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for field: field isCamelCase",
    },
  ],
}
`;
